### PR TITLE
Issuenumber5/#131/createrankinglist

### PR DESCRIPTION
**Pull request by Jeilef**

### CODE REVIEW

File: app/assets/javascripts/eventscripts/create_event.js
Comment (stlutz) [inline app/assets/javascripts/eventscripts/create_event.js]: I don't know any JavaScript, but this seems a bit ugly to me. As an idea: Might it be possible to define these fields as a collection and then map the enableField (or disableField) method onto each element?

File: app/assets/javascripts/eventscripts/create_event.js
Comment (Askir) [inline app/assets/javascripts/eventscripts/create_event.js]: Ooops, this is quite unfortunate, in the scope of #145 we decided to to quite the big refactor of the create progress, and now each event type has their own partial. So all the javascript isn't needed anymore. I talked about this with my PO @MarcelSchmidberger beforehand, so there must have been some misscomunication?

We'll have to figure out what the final call on this is.

File: app/helpers/events_helper.rb
Comment (Askir) [inline app/helpers/events_helper.rb:11]: Good job on these though for sure, missing localization is always a topic when talking to POs :D

File: app/models/event.rb
Comment (Askir) [inline app/models/event.rb]: These validations should just be in the according models if they aren't needed for a rankinglist. Just move them to the Tournament and League instead of having the extra check. Don't forget updating the Factories aswell though :)

File: app/models/rankinglist.rb
Comment (Askir) [inline app/models/rankinglist.rb]: Just a conceptual question: Does Playertype truely not matter here? Because if it doesn't we should maybe talk about removing the field all together for consitency sake. I'm not sure if a League and Tournament have to have this field. Haven't talked to POs about this yet.
Also why is discipline not needed for a ranking list? Isn't a rankinglist for just one discipline (like chess for example) aswell? Otherwise against we need to remove that one from the events validations and put all these validations into the right models.

File: app/models/event.rb
Comment (Askir) [inline app/models/event.rb]: Put this enum into the Rankinglist model. It's not really need for any other type of event.

File: app/models/event.rb
Comment (Askir) [inline app/models/event.rb]: This method is not needed if the validation is in the Rankinglist model class instead of here.

File: app/models/tournament.rb
Comment (Askir) [inline app/models/tournament.rb]: good job :) but add this to `league.rb` aswell please, if you end up removing it from `event.rb`

File: app/views/events/_form.html.erb
Comment (Askir) [inline app/views/events/_form.html.erb]: Again this whole file should look a lot cleaner in the way we refactored the model and the views in the process of #145.
You should just be able to create another partial file: `_rankinglist.html.erb` and add your additional fields there. If you open up the link `/events/new?type=rankinglist` it should automatically render the correct form.

File: app/models/event.rb
Comment (stlutz) [inline app/models/event.rb]: This would make more sense in the Tournament and League subclasses (where those fields are actually required fields).

File: db/migrate/20171208151635_add_metric_to_events.rb
Comment (Askir) [inline db/migrate/20171208151635_add_metric_to_events.rb]: game_mode is supposed to handle what you are calling metric right now the tournament saves its tournament type(KO, double-KO) in the column. The League saves the league type(round robin, double round robin, etc.) in that column, so I would suggest that the rankinglist saves what system it uses in that column aswell (Elo, trueskill, or whatever).

File: db/migrate/20171212084905_change_game_mode_and_player_types_null.rb
Comment (Askir) [inline db/migrate/20171212084905_change_game_mode_and_player_types_null.rb]: If you end up saving it in the same column, there is no need for this in the migration and also we obviously have to figure out if player_type is gonna stay or not.

File: app/models/event.rb
Comment (stlutz) [inline app/models/event.rb]: Checking for the class name could probably be avoided by (as mentioned before) moving the corresponding validation to the relevant subclasses.

File: app/models/rankinglist.rb
Comment (stlutz) [inline app/models/rankinglist.rb]: Why does the Rankinglist not want a discipline or a player_type? Doesn't Event already check for discipline's presence?

File: spec/factories/rankinglists.rb
Comment (Askir) [inline spec/factories/rankinglists.rb]: make this a child of event by using `parent: :event` and set the class to rankinglist by `class: :rankinglist` See the tournament and league factory on dev if you want an example. Also if you end up removing the start and enddate from the even factory itself (because it isn't needed for rankinglists) please add it into the tournament and league factories.

File: app/views/events/_form.html.erb
Comment (stlutz) [inline app/views/events/_form.html.erb]: We decided to use the colon notation in favor of the arrow notation for mapping key value pairs in hashes.

File: config/locales/de.yml
Comment (stlutz) [inline config/locales/de.yml]: Thank you.

**Review by Askir — CHANGES_REQUESTED** [review]
Sadly we actually have some conflicting information from our POs it seems. Which happens if mutliple teams work on similar pages. I hope we can clear this up.

However I also left some general info on the usage of the event model.
We changed how Events are being created in one of our latest commits, you should just be able to add the ranking list in the create method of the controller.
For more information on how to use the event model you can also look at the small article in the wiki that I created: https://github.com/hpi-swt2/sport-portal/wiki/Das-Event-Model

If there are any other questions on this, feel free to ask me. I can also help you merge this once we have a definitive answer from  our POs on how the create event page should actually look like.

File: config/locales/de.yml
Comment (stlutz) [inline config/locales/de.yml]: According to Wikipedia:

> Ein Runder Robin, englisch Round-robin, ist eine Bitt- und Beschwerdeschrift, auch Petition genannt, die von ihren Verfassern und Befürwortern rundherum in einem Kreis unterschrieben worden ist

I don't think that's what was meant. Other translation include "Jeder-gegen-Jeden" or "Rundenturnier". Even just keeping the english original "Round Robin" would have been more explanatory.
#19 states "Jeder-gegen-Jeden" to be the wanted translation.

File: config/locales/de.yml
Comment (stlutz) [inline config/locales/de.yml]: The correct plural of 'half' is 'halves'. The correct translation was sadly already defined in #19. Therefore it should be "Hin- und Rückrunde".

File: spec/factories/rankinglists.rb
Comment (stlutz) [inline spec/factories/rankinglists.rb:1]: Does this really output a valid Rankinglist? This does not look like it would pass Event's validations.

**Review by stlutz — CHANGES_REQUESTED** [review]
The Event's validation part seems off to me. Reintroduction of colon notation in events' _form has to be reverted. Translation needs to be according to #19.

File: app/models/rankinglist.rb
Comment (stlutz) [reply app/models/rankinglist.rb]: Btw. tests testing whether it is possible to actually create a Rankinglist would be really nice.

File: app/assets/javascripts/eventscripts/create_event.js
Comment (sumpfranze) [reply app/assets/javascripts/eventscripts/create_event.js]: We got the same from our PO, however, for the refactoring of the create Event process, a whole other User story #227 exists. So I would suggest to do that refactoring in the context of that User Story.

File: app/models/event.rb
Comment (sumpfranze) [reply app/models/event.rb]: :+1:

File: app/models/event.rb
Comment (sumpfranze) [reply app/models/event.rb]: I guess you could do it both ways. Moving it to the Tournament and League models you would get a bit of code duplication, this way you have to "disable" the check for ranking lists. But probably with how few subclasses we have, moving it to League and Tournament would be a bit cleaner

File: app/models/rankinglist.rb
Comment (sumpfranze) [reply app/models/rankinglist.rb]: I know that my PO Pius @GittiHab likes to have a selection in the create event form for single player/team events. That doesn't necessarily mean we need to store that in the database, though. If we start working with min players per team and max players per team fields, we could infer if it is a single player event, if min players and max players are 1. But yeah, that's something we need to reach a conclusion on.

Discipline should probably be available for ranking lists, though.

In general though, this whole line should be removed, since it doesn't matter what is in these columns anyway and the presence: true validation in the events model is already disabled for ranking lists (Can also change by moving it to the models).

File: app/views/events/_form.html.erb
Comment (sumpfranze) [reply app/views/events/_form.html.erb]: something that will probably be done in the process of #227

File: db/migrate/20171208151635_add_metric_to_events.rb
Comment (sumpfranze) [reply db/migrate/20171208151635_add_metric_to_events.rb]: that's a nice suggestion

File: spec/factories/rankinglists.rb
Comment (sumpfranze) [reply spec/factories/rankinglists.rb]: I dont think it is necessary to remove `:deadline`, `:enddate` and `:startdate` from the events factory, since you could also just set those values to nil in the ranking list values when they don't get validated.

File: app/views/events/_form.html.erb
Comment (Askir) [reply app/views/events/_form.html.erb]: The refactoring has already been done and is merged in dev. It's also been done by team swteam in one of their user stories, so #227 is basically obsolete if we just merge it correctly now.

File: spec/factories/rankinglists.rb
Comment (Askir) [reply spec/factories/rankinglists.rb]: Factories are supposed to create the minimal needed version of a model though, so if a rankinglist doesn't need these dates we should remove it from the model as necessary validation and therefore remove it from the factory aswell.
We could however add a subfactory :event_with_dates which has these values and then let :league and :tournament factories inherit from this factory instead of the base one. That way we don't have to set the values to nil in the rankinglist factory.

File: app/models/event.rb
Comment (stlutz) [inline app/models/event.rb]: Is there a reason this was not moved to the tournament and league subclasses together with the other validations?

Comment (Askir) [inline app/models/event.rb]: I think moving this validation and the one for max_teams to tournaments and leagues is the best we can do.  
Sure code duplication might not be super pretty but it's probably still cleaner than having a type check in the event model.

File: app/views/events/forms/_League.html.erb
Comment (Askir) [inline app/views/events/forms/_League.html.erb:1]: I don't think you have to put an absolute path here, just Dates should be enough.

File: spec/features/event/new_event_spec.rb
Comment (stlutz) [inline spec/features/event/new_event_spec.rb:85]: This test contains assumptions about the validity of events. Using a factory to fill in the fields could solve that.

File: spec/models/tournament_spec.rb
Comment (stlutz) [inline spec/models/tournament_spec.rb:26]: A similar test would be useful for ranking lists.

File: app/views/events/forms/_Tournament.html.erb
Comment (Askir) [inline app/views/events/forms/_Tournament.html.erb]: There are a few cases where oldschool hashing syntax is used with :key => "value" instead of key: "value". Maybe just let rubocop fix this if it's possible to use in views?

File: config/locales/de.yml
Comment (Askir) [inline config/locales/de.yml]: I would personally not translate true skill at all, since it's a specific way of calculating the ranking data and nobody knows what "Wahres Können" is representing.

File: db/migrate/20171212084905_change_game_mode_and_player_types_null.rb
Comment (Askir) [inline db/migrate/20171212084905_change_game_mode_and_player_types_null.rb]: I'm still a bit surprised that this isn't important for you guys at all. Single players can only join teams with the player_type :single and teams will only be able to join events with the player_type :team. So if you let this column be 0 the join and leave methods will not work for you anymore. If you want to add both you can just add that option to the enum and change the code in join and leave to permit both if the option is selected.

Otherwise noone will be able to join a rankinglist (maybe this is intended?)

File: spec/views/events/new.html.erb_spec.rb
Comment (Askir) [inline spec/views/events/new.html.erb_spec.rb:24]: I don't know what the current styling agreement is, because these two have just been switched in one of the latest PRs I think.

File: spec/factories/rankinglists.rb
Comment (Askir) [inline spec/factories/rankinglists.rb]: This is definitely not the cleanest solution to a good rankinglist factory. In my opinion. the :event factory should not specify all the dates and player_type (if you end up actually not using it), but instead have a subfactory `:event_wtih_dates`
that specifies all these values. That way you can inherit from the base factory and can cut out the lines 5-8. To make tournaments and leagues work again they just have to inherit from `:event_with_dates.`

**Review by Askir — CHANGES_REQUESTED** [review]
The factory and validation should be changed before merging this. Also please check if you really don't want to use player_type in a rankinglist. Because this will disable the joining of single players AND teams. So I'm not sure if that is intended.

Otherwise: great job on the merge! I definitely thought I'd have to comment on more than this :D

File: spec/factories/rankinglists.rb
Comment (sumpfranze) [reply spec/factories/rankinglists.rb]: Well, if the dates are required in the events model, but not in the ranking list model, then setting them in the event factory and setting them to nil in the ranking factory would be the best option in my opinion. That way there are only required values initialized by the factory and the others are nil.

If the validations get moved to the submodels, then I agree, the factories should also display that movement.

**Review by stlutz — APPROVED** [review]

**Comment by Askir** [comment]
Codeclimate issue is intended to let people, who implement subclasses of event know what to expect as variable.

**Review by Askir — APPROVED** [review]

**Pull request merged**
